src/refract/JsonValue.cc

```diff
@@ -0,0 +1,499 @@
+
+            if (const auto& sampleValue = findSample(e)) {
+                return renderValue(*sampleValue, options);
+            }
+
+            if (const auto& defaultValue = findDefault(e)) {
+                return renderValue(*defaultValue, options);
+            }
+
+            LOG(warning) << "no value for non-nullable empty StringElement; using `\"\"`";
+            return so::String{};
+        }
+
+        return so::String{ e.get().get() };
+    }
+
+    so::Value renderValue(const NumberElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering NumberElement to JSON Value";
+
+        options = updateTypeAttributes(e, options);
+
+        if (e.empty()) {
+            if (options.test(NULLABLE_FLAG))
+                return so::Null{};
+
+            if (const auto& sampleValue = findSample(e)) {
+                return renderValue(*sampleValue, options);
+            }
+
+            if (const auto& defaultValue = findDefault(e)) {
+                return renderValue(*defaultValue, options);
+            }
+
+            LOG(warning) << "no value for non-nullable empty NumberElement; using `0`";
+            return so::Number{};
+        }
+
+        return so::Number{ e.get().get() };
+    }
+
+    so::Value renderValue(const BooleanElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering BooleanElement to JSON Value";
+
+        options = updateTypeAttributes(e, options);
+
+        if (e.empty()) {
+            if (options.test(NULLABLE_FLAG))
+                return so::Null{};
+
+            if (const auto& sampleValue = findSample(e)) {
+                return renderValue(*sampleValue, options);
+            }
+
+            if (const auto& defaultValue = findDefault(e)) {
+                return renderValue(*defaultValue, options);
+            }
+
+            LOG(warning) << "no value for non-nullable empty NumberElement; using `false`";
+            return so::False{};
+        }
+
+        if (e.get().get())
+            return so::True{};
+        return so::False{};
+    }
+
+    so::Value renderValue(const ExtendElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering ExtendElement to JSON Value";
+
+        auto merged = e.get().merge();
+        assert(merged);
+        return renderValue(*merged, options);
+    }
+
+} // namespace
+
+namespace
+{
+
+    void renderProperty(so::Object& obj, const MemberElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering property MemberElement as JSON Value";
+
+        if (hasFixedTypeAttr(e))
+            options.set(FIXED_FLAG);
+
+        options.set(FIXED_TYPE_FLAG, hasFixedTypeTypeAttr(e));
+        options.set(NULLABLE_FLAG, hasNullableTypeAttr(e));
+
+        if (hasRequiredTypeAttr(e))
+            options.set(REQUIRED_FLAG);
+
+        if (hasOptionalTypeAttr(e))
+            options.reset(REQUIRED_FLAG);
+
+        const auto* k = e.get().key();
+        const auto* v = e.get().value();
```

Comment:
IMO would be nicer to use clearer variable names instead of `k`, `v`, `e`, `s`, `a`. Descriptive variable names go a long way in making this more readable than trying to remember which character represents which things. We're humans not machines, readability counts.

Reply:
Agreed